File: Assignment 5/Go/SplitFileServer.go
/**
 * SplitFileServer.go
 * ID : 20194094
 * Name : Yongmin Yoo
 **/

package main

import (
	"fmt"
	"io"
	"net"
	"os"
	"os/signal"
	"path/filepath"
	"strconv"
	"strings"
	"syscall"
)

var serverPort int
var filenameSuffix string

func main() {
	if len(os.Args) != 2 {
		exitError("Usage: go run server.go <port>")
	}

	serverPortArgument, err := strconv.Atoi(os.Args[1])
	if err != nil {
		exitError("Invalid Argument")
	}

	serverPort = serverPortArgument
	if serverPort/10000 == 4 {
		filenameSuffix = "-part1"
	} else {
		filenameSuffix = "-part2"
	}

	serverListener := initServer()

	sigintHandler := make(chan os.Signal, 1)
	signal.Notify(sigintHandler, syscall.SIGINT)
	go func() {
		<-sigintHandler
		closeServer(serverListener)
		os.Exit(0)
	}()

	for {
		if serverListener != nil {
			conn, _ := serverListener.Accept()
			go handleConnection(conn)
		}
	}
}

func closeServer(serverConn net.Listener) {
	fmt.Println("\rClosing Server Program...\nBye bye~")
	serverConn.Close()
}

func initServer() net.Listener {
	serverListener, err := net.Listen("tcp4", fmt.Sprintf(":%d", serverPort))
	if err != nil {
		exitError(err.Error())
	}
	fmt.Println("Server listening on port", serverPort)

	return serverListener
}

func handleConnection(conn net.Conn) {
	defer func() {
		if conn != nil {
			conn.Close()
		}
	}()

	readBuffer := make([]byte, 1024)
	for {
		if conn == nil {
			return
		}
		readLength, err := conn.Read(readBuffer)
		if err != nil && err != io.EOF {
			fmt.Println("Error reading:", err.Error())
			return
		}
		if readLength == 0 {
			return
		}

		readData := string(readBuffer[:readLength])
		readDataParts := strings.SplitN(readData, ":", 2)
		if len(readDataParts) != 2 {
			fmt.Println("Invalid message format")
			return
		}

		cmd, filename := readDataParts[0], readDataParts[1]
		if cmd == "PUT" {
			conn.Write([]byte("READY\n"))
			saveHalfFile(conn, filename)
		} else if cmd == "GET" {
			sendHalfFile(conn, filename)
		} else {
			fmt.Println("Unknown command")
			return
		}
	}
}

func saveHalfFile(conn net.Conn, filename string) {
	partFilename := getPartedFilename(filename)
	partFile, err := os.Create(partFilename)
	if err != nil {
		fmt.Println("Error creating file:", err.Error())
		conn.Write([]byte("ERROR\n"))
		return
	}
	defer partFile.Close()

	partFileBuffer := make([]byte, 1024)
	for {
		partFileLength, err := conn.Read(partFileBuffer)
		if err != nil && err != io.EOF {
			fmt.Println("Error reading:", err.Error())
			conn.Write([]byte("ERROR\n"))
			return
		}
		if partFileLength == 0 {
			return
		}
		if _, err := partFile.Write(partFileBuffer[:partFileLength]); err != nil {
			fmt.Println("Error writing to file:", err.Error())
			conn.Write([]byte("ERROR\n"))
			return
		}
	}
}

func sendHalfFile(conn net.Conn, filename string) {
	partFilename := getPartedFilename(filename)
	partFile, err := os.Open(partFilename)
	if err != nil {
		fmt.Println("Error opening file:", err.Error())
		conn.Write([]byte("ERROR\n"))
		return
	}
	defer partFile.Close()

	partFileBuffer := make([]byte, 1024)
	for {
		partFileLength, err := partFile.Read(partFileBuffer)
		if err != nil {
			if err == io.EOF {
				break
			}

			fmt.Println("Error reading file:", err.Error())
			conn.Write([]byte("ERROR\n"))
			return
		}
		if partFileLength == 0 {
			break
		}

		conn.Write(append([]byte("N"), partFileBuffer[:partFileLength]...))
	}

	conn.Write([]byte("EOF"))
}

func getPartedFilename(filename string) string {
	ext := filepath.Ext(filename)
	base := strings.TrimSuffix(filename, ext)
	return base + filenameSuffix + ext
}

func exitError(msg string) {
	_, err := fmt.Fprintf(os.Stderr, "Error: %s\n", msg)
	if err != nil {
		fmt.Printf("Error: %s\n", msg)
	}
	os.Exit(1)
}

File: Assignment 5/Java/SplitFileServer.java
import java.io.*;
import java.net.*;
import java.nio.file.*;
import java.util.concurrent.*;

public class SplitFileServer {
    private static int serverPort;
    private static String filenameSuffix;

    public static void main(String[] args) {
        if (args.length != 1) {
            exitError("Usage: java SplitFileServer <port>");
        }

        try {
            serverPort = Integer.parseInt(args[0]);
        } catch (NumberFormatException e) {
            exitError("Invalid Argument");
        }

        if (serverPort / 10000 == 4) {
            filenameSuffix = "-part1";
        } else {
            filenameSuffix = "-part2";
        }

        ServerSocket serverSocket = initServer();

        Runtime.getRuntime().addShutdownHook(new Thread(() -> closeServer(serverSocket)));

        ExecutorService executorService = Executors.newCachedThreadPool();
        while (true) {
            try {
                Socket socket = serverSocket.accept();
                executorService.execute(() -> handleConnection(socket));
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    private static void closeServer(ServerSocket serverSocket) {
        System.out.println("\rClosing Server Program...\nBye bye~");
        try {
            if (serverSocket != null && !serverSocket.isClosed()) {
                serverSocket.close();
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private static ServerSocket initServer() {
        try {
            ServerSocket serverSocket = new ServerSocket(serverPort);
            System.out.println("Server listening on port " + serverPort);
            return serverSocket;
        } catch (IOException e) {
            exitError(e.getMessage());
            return null;
        }
    }

    private static void handleConnection(Socket socket) {
        try (BufferedReader reader = new BufferedReader(new InputStreamReader(socket.getInputStream()));
             PrintWriter writer = new PrintWriter(socket.getOutputStream(), true)) {

            char[] readBuffer = new char[1024];
            int readLength = reader.read(readBuffer);
            if (readLength == -1) {
                return;
            }

            String readData = new String(readBuffer, 0, readLength);
            String[] readDataParts = readData.split(":", 2);
            if (readDataParts.length != 2) {
                System.out.println("Invalid message format");
                return;
            }

            String cmd = readDataParts[0];
            String filename = readDataParts[1];
            if ("PUT".equals(cmd)) {
                writer.println("READY");
                saveHalfFile(socket, filename);
            } else if ("GET".equals(cmd)) {
                sendHalfFile(socket, filename);
            } else {
                System.out.println("Unknown command");
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    private static void saveHalfFile(Socket socket, String filename) {
        String partFilename = getPartedFilename(filename);
        try (FileOutputStream partFile = new FileOutputStream(partFilename);
             InputStream socketInput = socket.getInputStream()) {

            byte[] partFileBuffer = new byte[1024];
            int partFileLength;
            while ((partFileLength = socketInput.read(partFileBuffer)) != -1) {
                partFile.write(partFileBuffer, 0, partFileLength);
            }
        } catch (IOException e) {
            System.out.println("Error handling file: " + e.getMessage());
        }
    }

    private static void sendHalfFile(Socket socket, String filename) {
        String partFilename = getPartedFilename(filename);
        try (FileInputStream partFile = new FileInputStream(partFilename);
             OutputStream socketOutput = socket.getOutputStream()) {

            byte[] partFileBuffer = new byte[1024];
            int partFileLength;
            while ((partFileLength = partFile.read(partFileBuffer)) != -1) {
                byte[] prefixedBuffer = new byte[partFileLength + 1];
                prefixedBuffer[0] = 'N';
                System.arraycopy(partFileBuffer, 0, prefixedBuffer, 1, partFileLength);
                socketOutput.write(prefixedBuffer, 0, prefixedBuffer.length);
            }
            socketOutput.write("EOF".getBytes());
        } catch (FileNotFoundException e) {
            System.out.println("Error opening file: " + e.getMessage());
        } catch (IOException e) {
            System.out.println("Error handling file: " + e.getMessage());
        }
    }

    private static String getPartedFilename(String filename) {
        String ext = filename.substring(filename.lastIndexOf('.'));
        String base = filename.substring(0, filename.lastIndexOf('.'));
        return base + filenameSuffix + ext;
    }

    private static void exitError(String msg) {
        System.err.println("Error: " + msg);
        System.exit(1);
    }
}
